front/lib/resources/membership_resource.ts

```diff
@@ -0,0 +1,344 @@
+  ModelStatic,
+  Transaction,
+  WhereOptions,
+} from "sequelize";
+import { Op } from "sequelize";
+
+import type { Workspace } from "@app/lib/models";
+import { BaseResource } from "@app/lib/resources/base_resource";
+import { MembershipModel } from "@app/lib/resources/storage/models/membership";
+import type { ReadonlyAttributesType } from "@app/lib/resources/storage/types";
+import logger from "@app/logger/logger";
+
+type GetMembershipsOptions = RequireAtLeastOne<{
+  userIds: number[];
+  workspace: WorkspaceType | Workspace;
+}> & {
+  roles?: MembershipRoleType[];
+};
+
+// Attributes are marked as read-only to reflect the stateless nature of our Resource.
+// This design will be moved up to BaseResource once we transition away from Sequelize.
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface MembershipResource
+  extends ReadonlyAttributesType<MembershipModel> {}
+export class MembershipResource extends BaseResource<MembershipModel> {
+  static model: ModelStatic<MembershipModel> = MembershipModel;
+
+  constructor(
+    model: ModelStatic<MembershipModel>,
+    blob: Attributes<MembershipModel>
+  ) {
+    super(MembershipModel, blob);
+  }
+
+  static async getActiveMemberships({
+    userIds,
+    workspace,
+    roles,
+  }: GetMembershipsOptions): Promise<MembershipResource[]> {
+    const whereClause: WhereOptions<InferAttributes<MembershipModel>> = {
+      startAt: {
+        [Op.lte]: new Date(),
+      },
+      endAt: {
+        [Op.or]: [{ [Op.eq]: null }, { [Op.gte]: new Date() }],
+      },
+    };
+
+    if (userIds) {
+      whereClause.userId = userIds;
+    }
+    if (workspace) {
+      whereClause.workspaceId = workspace.id;
+    }
+    if (roles) {
+      whereClause.role = {
+        [Op.in]: roles,
+      };
+    }
+
+    const memberships = await MembershipModel.findAll({
+      where: whereClause,
+    });
+
+    return memberships.map(
+      (membership) => new MembershipResource(MembershipModel, membership.get())
+    );
+  }
+
+  static async getLatestMemberships({
+    userIds,
+    workspace,
+    roles,
+  }: GetMembershipsOptions): Promise<MembershipResource[]> {
+    const orderedResourcesFromModels = (resources: MembershipModel[]) =>
+      resources
+        .sort((a, b) => a.startAt.getTime() - b.startAt.getTime())
+        .map(
+          (resource) => new MembershipResource(MembershipModel, resource.get())
+        );
+
+    const where: WhereOptions<InferAttributes<MembershipModel>> = {};
+    if (roles) {
+      where.role = {
+        [Op.in]: roles,
+      };
+    }
+    if (userIds) {
+      where.userId = {
+        [Op.in]: userIds,
+      };
+    }
+    if (workspace) {
+      where.workspaceId = workspace.id;
+    }
+
+    if (!workspace && !userIds?.length) {
+      throw new Error("At least one of workspace or userIds must be provided.");
+    }
+    if (userIds && !userIds.length) return [];
+
+    // Get all the memberships matching the criteria.
```

Comment:
Here we may be doing a bit of over-fetching if some members were revoked and re-invited several times. I believe this is fine since:
- the number of over-fetched stuff will be quite low
- it is not on a hot path 
- it makes the code / querying a lot simpler

Reply:
This is a hot path since it's called from `getLatestMembershipOfUserInWorkspace` which is called from login ?

Reply:
I think the over-fetch is fine, but noting this is quite hot

Reply:
Actually you don't get revoked memberships here do you?

Reply:
Yes we do get the revoked memberships, it's this method's purpose vs getActiveMemberships.
The overfetching comes from getting the non-latest memberships, which IMO is fine.